### PR TITLE
Added is_valid_email_address.rs

### PR DESCRIPTION
**Pull request by lyqio**
# Changes

- add is_valid_email_address.rs

## Type of change

- [x] New feature (non-breaking change which adds functionality)

## Checklist:

- [x] I ran bellow commands using the latest version of **rust nightly**.
- [x] I ran `cargo clippy --all -- -D warnings` just before my last commit and fixed any issue that was found.
- [x] I ran `cargo fmt` just before my last commit.
- [x] I ran `cargo test` just before my last commit and all tests passed.
- [x] I added my algorithm to the corresponding `mod.rs` file within its own folder, and in any parent folder(s).
- [x] I added my algorithm to `DIRECTORY.md` with the correct link.
- [x] I checked `COUNTRIBUTING.md` and my code follows its guidelines.

### CODE REVIEW

**Review by vil02 — CHANGES_REQUESTED** [review]
The definition of the _valid_ email address is slightly more involved, cf. https://en.wikipedia.org/wiki/Email_address#Examples
in particular, `"john..doe"@example.org` is perfectly fine.

This can be merged only if all of the _edge cases_ are covered.

**Review by lyqio — COMMENTED** [review]
I rewrote some of the code and now passes all of the tests on https://en.wikipedia.org/wiki/Email_address#Examples

File: src/string/is_valid_email_address.rs [hunk 1]
```diff
@@ -0,0 +1,139 @@
...
+    if domain.starts_with('.') || domain.ends_with('.') || domain.contains("..") {
+        return false;
+    }
+
+    // (8)  validate the dashes in the domain
+    if domain.starts_with('-') || domain.ends_with('-') {
+        return false;
+    }
+
+    true
```
Comment (vil02) [inline src/string/is_valid_email_address.rs:91]: What do you think about introducing functions like
- `is_local_part_valid`, which would then split into `is_quoted_local_part_valid` and `is_unquoted_local_part_valid`,
- `is_domain_valid`, which would then split accordingly.

File: src/string/is_valid_email_address.rs [hunk 2]
```diff
@@ -0,0 +1,139 @@
...
+        slashes: ("name/surname@example.com", true),
+        no_tld: ("admin@example", true),
+        tld: ("example@s.example", true),
+        quotes_with_space: ("\" \"@example.org", true),
+        quoted_double_dot: ("\"john..doe\"@example.org", true),
+        host_route: ("mailhost!username@example.org", true),
+        quoted_non_letters: (r#""very.(),:;<>[]\".VERY.\"very@\\ \"very\".unusual"@strange.example.com"#, true),
+        percent_symbol: ("user%example.com@example.org", true),
+        local_end_symbol: ("user-@example.org", true),
+        ip_address: ("postmaster@[123.123.123.123]", true),
```
Comment (vil02) [inline src/string/is_valid_email_address.rs:126]:
```suggestion
        ip_address: ("postmaster@[123.123.123.123]", true),
        wrong_ip_address: ("postmaster@[1234.123.123.123]", false),
```

File: src/string/is_valid_email_address.rs
Comment (vil02) [inline src/string/is_valid_email_address.rs:1]: Please make sure that _tests cover all of the lines_.

**Review by vil02 — CHANGES_REQUESTED** [review]

File: src/string/is_valid_email_address.rs (hunk 2 above)
Comment (lyqio) [reply src/string/is_valid_email_address.rs:196]: I added checks for ip addresses and added some extra tests to ensure that the ip checks work as expected.

File: src/string/is_valid_email_address.rs (hunk 1 above)
Comment (lyqio) [reply src/string/is_valid_email_address.rs]: Due to the way the code is written, checking the quoted part and unquoted part separately wouldn't make the code more readable, so instead I just refactored the code and split each piece of logic into smaller sections.

**Comment by lyqio** [comment]
I handled all the things that you requested.

File: src/string/is_valid_email_address.rs [hunk 3]
```diff
@@ -0,0 +1,213 @@
...
+        for section in sections {
+            let value = section.parse::<u32>();
+            if value.is_err() || value.unwrap() > 255 {
+                return false;
+            }
+        }
+    }
+
+    true
+}
```
Comment (vil02) [inline src/string/is_valid_email_address.rs:90]: Why not to split it into two functions: `is_ipv4` and `is_ipv6`? This will allow for testing them directly (currently lots of testes are missing e.g. `123.123.123.123.123.123.123.123` is not a valid ip address).

Furthermore, the logic here is quite convoluted. What do you think about checking (in `is_domain_valid`) if the _domain part_ has the form `[<some_str>]`, and if so, send the `<some_str>` to `is_ip` function. It could look like
```rust
fn is_valid_ip(ip_str: &str) -> bool {
    is_ipv4(ip_str) || is_ipv6(ip_str)
}
```

Such approach would me more natural: what is a valid ip address `255.255.255.255` or `[255.255.255.255]`?

**Review by vil02 — CHANGES_REQUESTED** [review]

**Comment by vil02** [comment]
Please make sure to update your branch (cf. #789) - otherwise the CI will fail.

**Comment by lyqio** [comment]
I think I fixed all the issues and made the IP checking better.

File: src/string/is_valid_email_address.rs [hunk 4]
```diff
@@ -0,0 +1,253 @@
...
+        if !domain.ends_with(']') {
+            return false;
+        }
+
+        let ip = &domain[1..domain.len() - 1];
+        return is_valid_ipv4(ip) || is_valid_ipv6(ip);
+    }
+
+    true
+}
```
Comment (vil02) [inline src/string/is_valid_email_address.rs:121]: This is quite convoluted and incorrect. Please follow the [earlier suggestions](https://github.com/TheAlgorithms/Rust/pull/783#discussion_r1759347825).

Note: `is_valid_email_address("postmaster@[")` returns `true`.

Also: `is_valid_ip("[]")` and `is_valid_ip("a")` would also return `true`.

**Review by vil02 — CHANGES_REQUESTED** [review]

File: src/string/is_valid_email_address.rs (hunk 4 above)
Comment (lyqio) [reply src/string/is_valid_email_address.rs]: I fixed the code and changed the is_valid_ip to make it fit your suggestion, It now passes the tests you provided.

File: src/string/is_valid_email_address.rs [hunk 5]
```diff
@@ -0,0 +1,255 @@
...
+
+        for chr in value.chars() {
+            if !HEX_CHARACTERS.contains(chr.to_ascii_lowercase()) {
+                return false;
+            }
+        }
+    }
+
+    true
+}
```
Comment (vil02) [inline src/string/is_valid_email_address.rs:104]: What do you think about testing these functions directly?

File: src/string/is_valid_email_address.rs [hunk 6]
```diff
@@ -0,0 +1,255 @@
...
+    {
+        return false;
+    }
+
+    if domain.starts_with('-') || domain.ends_with('-') {
+        return false;
+    }
+
+    true
+}
```
Comment (vil02) [inline src/string/is_valid_email_address.rs:164]: This could be simplified to
```rust
fn is_domain_valid(domain: &str) -> bool {
    if domain.starts_with('[') && domain.ends_with(']')
    {
        return is_valid_ip(&domain[1..domain.len() - 1])
    }

    if domain.len() > MAX_DOMAIN_LENGTH as usize {
        return false;
    }

    if !contains_legal_domain_characters(domain) {
        return false;
    }

    if has_bad_dots(domain) {
        return false;
    }

    if domain.starts_with('[') || domain.ends_with(']') {
        return false
    }

    if domain.starts_with('-') || domain.ends_with('-') {
        return false;
    }

    true
}
```
But it can be simplified even further: for example when you know that you are not dealing with the ip address, you know that a valid domain will not contain any bracket.

**Review by vil02 — CHANGES_REQUESTED** [review]
Please consider using more consistent naming: sometime the `valid` part is in the middle and sometimes at the end. Perhaps it is not needed at all?

File: src/string/is_valid_email_address.rs (hunk 5 above)
Comment (lyqio) [reply src/string/is_valid_email_address.rs]: I renamed `is_valid_ipv4` and `is_valid_ipv6` to `is_ipv4` and `is_ipv6` respectively, but for the rest of the functions which contained the word 'valid', I moved the valid to the middle of the function name to make the code more consistent while staying readable.

In addition I added tests for `is_ipv4` and `is_ipv6`, note that some ips were taken from the `is_valid_email_address` tests but I still kept those tests in as I thought they may still be good tests to have, they can be removed if you want. Also, due to adding more tests, I made two new macros, but the way I did it may not be the most efficient way of doing so, but it shouldn't matter too much as it is just for the tests.

File: src/string/is_valid_email_address.rs (hunk 6 above)
Comment (lyqio) [reply src/string/is_valid_email_address.rs]: I simplified the code as previously mentioned

File: src/string/is_valid_email_address.rs [hunk 7]
```diff
@@ -0,0 +1,303 @@
...
+    if !contains_legal_local_characters(local_part) {
+        return false;
+    }
+
+    if has_bad_dots(local_part) {
+        return false;
+    }
+
+    true
+}
```
Comment (vil02) [inline src/string/is_valid_email_address.rs:128]: What do you think about having a similar approach as in `is_valid_domain`? I.e. if the `local_part` starts and ends `"`. then a function `is_valid_local_quoted_part` and `is_valid_local_unquoted_part`?

Comment (lyqio) [reply src/string/is_valid_email_address.rs:128]: I don't think this would work as not all email addresses are made up of only a single string, for example `a."b".c` is valid although it does not begin and end with quotes. In addition due to the fact that the contents of all strings are removed at the start of the local part checking, I believe that it would make the code less readable.

Comment (vil02) [reply src/string/is_valid_email_address.rs:128]: `a."b".c` is _unquoted_. It contains a `"` character, therefore it is not valid (cf. [wiki](https://en.wikipedia.org/wiki/Email_address#Local-part)).
